### PR TITLE
add PdfImageAsset

### PR DESCRIPTION
**Pull request by ArielMejiaDev**
# What makes the PR

It adds a new class helper that encodes as base 64 images, no matter if the image path is relative or an absolute path, and also adds some prefix required for image assets in headers and footers to display images successfully by puppeteer.

The code would look like this

```php
    public function __invoke(): PdfBuilder
    {
        return pdf('pdf.invoice', [
            'invoiceNumber' => '1234',
            'customerName' => 'Grumpy Cat',
        ])->footerView('components.footer', [
            'image' => PdfImageAsset::make('https://avatars.githubusercontent.com/u/7535935?s=200&v=4')
        ])->headerView('components.header', [
            'image' => PdfImageAsset::make(public_path('assets/spatie-logo.png')),
        ]);
    }
```

It would solve a lot of discussions and issues about how to add images to headers and footers.

like: 
- #46 
- #67 
- #69 

I add it as a support class as its concern is simple, I doubt adding the functionality as a trait to the PdfBuilder class but just makes more sense to handle it as a static helper.

## Testing

It also adds its test `PdfImageAssetTest.

## Docs

It also adds some suggested text for docs in headers and footers.

## Demo about the issue

It solves the issue talked about in discussion #71 

https://www.loom.com/share/8912fba845214a1d93aefc99fcf93b93?sid=c0389007-fcdf-4110-bee1-82317fc53c91

### CODE REVIEW

**Comment by freekmurze** [comment]
Thanks for your work on this PR.

I like the functionality, but I think it could be make easier to use. Instead of letting users call `PdfImageAsset::make` in a controller, let's provide a Blade directive for them.

It could work like this in a Blade view

```
@inlidedImage($url)
```

This way, the controller doesn't have to know that images should be inlined.

Could you make this change? Thanks!

**Comment by RVP04** [comment]
@freekmurze appreciated for the suggestion @freekmurze welcoming to do it and waiting for the push

**Comment by ArielMejiaDev** [comment]
> Thanks for your work on this PR.
> 
> I like the functionality, but I think it could be make easier to use. Instead of letting users call `PdfImageAsset::make` in a controller, let's provide a Blade directive for them.
> 
> It could work like this in a Blade view
> 
> ```
> @inlidedImage($url)
> ```
> 
> This way, the controller doesn't have to know that images should be inlined.
> 
> Could you make this change? Thanks!

I am on it! thanks

**Comment by RVP04** [comment]
Eagerly waiting @ArielMejiaDev

**Comment by ArielMejiaDev** [comment]
Hi @freekmurze there is a commit, if you can check this loom (I try to make it short & fun, sorry the audio) you would see the behavior now with a directive (looks lot more cleaner).

## Demo

https://www.loom.com/share/e3a5fc63f7894b779371e4c5d13a6a0d?sid=80074810-3d33-4ac6-9293-9d7061afc2b4

I was not able to run all Pest tests that uses `toContainText` because I do not find how to install `pdfToText` package with Hombrew

```php
expect()->extend('toContainText', function (string|array $expectedText) {
    $binPath = PHP_OS === 'Linux'
        ? '/usr/bin/pdftotext'
        : '/opt/homebrew/bin/pdftotext';
```

**Comment by RVP04** [comment]
@freekmurze waiting for you

**Comment by RVP04** [comment]
@freekmurze waiting please

**Comment by freekmurze** [comment]
@RVP04 Thanks for your enthusiasm but please refrain from posting reminders here and on other social channels. That won't make me working faster.

File: docs/basic-usage/formatting-pdfs.md [hunk 1]
```diff
@@ -48,6 +48,16 @@ Inside the footer, you can use the following Blade directives:
 - `@pageNumber`:  The current page number
 - `@totalPages`:  The total number of pages
 
+### Display Images in Headers and Footers
+
+You can add an image using the blade directive `@inlidedImage`
```
Comment (freekmurze) [inline docs/basic-usage/formatting-pdfs.md:53]: This should be renamed to `inlinedImage` everywhere in this PR.

**Review by freekmurze — CHANGES_REQUESTED** [review]
@ArielMejiaDev Thanks for your work on this!

**Comment by RVP04** [comment]
Because I am eagerly waiting for you

**Comment by freekmurze** [comment]
Thanks!

**Comment by Professor93M** [comment]
How I can use **PdfImageAsset** with the pdf. because it's not found?
and I tried to add image to the header but it's not worked

**Comment by ArielMejiaDev** [comment]
> How I can use **PdfImageAsset** with the pdf. because it's not found? and I tried to add image to the header but it's not worked

@Professor93M 

You can set images using the directive:

```php
@imageInlined('your-image.png')
```

**Comment by RVP04** [comment]
You have to use 

'''
@imageInlined('your-image.png')
'''

Check yours it is wrong

**Comment by ArielMejiaDev** [comment]
Hi @Professor93M as @RVP04 says sorry the correct directive is `@imageInlined()`, so your current code should look like this:

```php
@imageInlined('images/logo.png')
```

or using an object property

```php
@imageInlined($model->image)
```

**Comment by Professor93M** [comment]
Do you mean like that:

```
<div dir="rtl" style="miargin: 20px auto; padding: 20px 70px; width: 100%;">
    <div style="display: flex; justify-content: space-between; align-items: center;">
        <h1 style="font-size: 16px; font-weight: 700; width: 150px; text-align: center;">Title</h1>
        @imageInlined('images/logo.png')
        <h1 style="font-size: 14px; font-weight: 700; width: 200px; text-align: center;">Title2</h1>
    </div>
    <hr style="border: 1px solid #DE9C55; width: 100%;">
</div>
```

```
return pdf('Print.pdf', compact('invoice', 'exchange'))
            ->footerView('Print.footer')
            ->headerView('Print.header')
            ->margins(30, 5, 15, 5)
            ->landscape()
            ->format(Format::A5);
```

I tried that but it show only the text like that: **@imageInlined('images/logo.png')**

**Comment by ArielMejiaDev** [comment]
@Professor93M please check first that the file exists in your public directory, then you can follow this example https://github.com/ArielMejiaDev/laravel-pdf-demo/blob/main/resources/views/components/header.blade.php you can check the whole app to see controllers, public directory, resources/views to check blade files etc.

**Comment by Professor93M** [comment]
I used **@imageInlined** directive but it's not work with me.
I used this code I think it's work but there is problem with my server I think
```
<figure>
    @inlinedImage('images/logo.png')
</figure>
```

It show me this page even if I refresh it:
image

and I have to stop the server and restart it again.

**Comment by RVP04** [comment]
First check in view whether file is loading or not. Share your actual code controller and view properly to troubleshoot

**Comment by Professor93M** [comment]
This my code:

**Controller**
```
public function invoice($id)
    {
        return pdf('Print.pdf')
            ->footerView('Print.footer')
            ->headerView('Print.header')
            ->margins(30, 5, 15, 5)
            ->landscape()
            ->format(Format::A5);
    }
```

**pdf.blade.php (View)**
```
<html>
    <head>
        <title>Invoice</title>
        <script src="https://cdn.tailwindcss.com"></script>
    </head>
    <body dir="rtl">
        <div class="flex mx-auto space-x-2 sm:space-x-4 font-['Almarai']">
            <div class="w-full px-10">
                <h1 class="text-xl text-center font-bold">Invoice</h1>
            </div>
        </div>
    </body>
</html>
```

**header.blade.php** 
```
<div dir="rtl" style="miargin: 20px auto; padding: 20px 70px; width: 100%;">
    <div style="display: flex; justify-content: space-between; align-items: center;">
        <h1 style="font-size: 16px; font-weight: 700; width: 150px; text-align: center;">Title</h1>
        <figure>
            @inlinedImage('images/logo.png')
        </figure>
        <h1 style="font-size: 14px; font-weight: 700; width: 200px; text-align: center;">Title2</h1>
    </div>
    <hr style="border: 1px solid #DE9C55; width: 100%;">
</div>
```

I run `php artisan optimize`

same problem

**Comment by Professor93M** [comment]
Any comment gents?
Please need help

**Comment by RVP04** [comment]
Try to trouble shoot with views first it is getting loaded or not then you we decide for the solution

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by amirnbcode** [comment]
still the same problem not showing in header 
I wen trough all mentioned ways above 
but not working

**Comment by A909M** [comment]
> This my code:
> 
> **Controller**
> 
> ```
> public function invoice($id)
>     {
>         return pdf('Print.pdf')
>             ->footerView('Print.footer')
>             ->headerView('Print.header')
>             ->margins(30, 5, 15, 5)
>             ->landscape()
>             ->format(Format::A5);
>     }
> ```
> 
> **pdf.blade.php (View)**
> 
> ```
> <html>
>     <head>
>         <title>Invoice</title>
>         <script src="https://cdn.tailwindcss.com"></script>
>     </head>
>     <body dir="rtl">
>         <div class="flex mx-auto space-x-2 sm:space-x-4 font-['Almarai']">
>             <div class="w-full px-10">
>                 <h1 class="text-xl text-center font-bold">Invoice</h1>
>             </div>
>         </div>
>     </body>
> </html>
> ```
> 
> **header.blade.php**
> 
> ```
> <div dir="rtl" style="miargin: 20px auto; padding: 20px 70px; width: 100%;">
>     <div style="display: flex; justify-content: space-between; align-items: center;">
>         <h1 style="font-size: 16px; font-weight: 700; width: 150px; text-align: center;">Title</h1>
>         <figure>
>             @inlinedImage('images/logo.png')
>         </figure>
>         <h1 style="font-size: 14px; font-weight: 700; width: 200px; text-align: center;">Title2</h1>
>     </div>
>     <hr style="border: 1px solid #DE9C55; width: 100%;">
> </div>
> ```
> 
> I run `php artisan optimize`
> 
> same problem

This  [comment ](https://github.com/puppeteer/puppeteer/issues/2604#issuecomment-2317217362) solve my problem.

just add 
`->withBrowsershot(function (Browsershot $browsershot) {
        return $browsershot->writeOptionsToFile(); 
    })`
 [see the document ](https://spatie.be/docs/browsershot/v4/miscellaneous-options/writing-options-to-a-file) 
 
 your code will be like this:.

```
public function invoice($id)
    {
        return pdf('Print.pdf')
            ->footerView('Print.footer')
            ->headerView('Print.header')
            ->margins(30, 5, 15, 5)
            ->landscape()
            ->withBrowsershot(function (Browsershot $browsershot) {
                   return $browsershot->writeOptionsToFile(); 
              })
            ->format(Format::A5);
    }
```

**Pull request merged**